draft-ietf-dnsop-svcb-httpssvc.md

```diff
@@ -743,38 +745,49 @@ responses to the address queries that were issued in parallel.
 A few initial SvcParamKeys are defined here.  These keys are useful for
 HTTPS, and most are applicable to other protocols as well.
 
-## "alpn"
-
-The "alpn" SvcParamKey defines the Application Layer Protocol
-(ALPN, as defined in {{!RFC7301}) supported by a TLS-based alternative
-service.  Its value SHOULD be an entry in the IANA registry "TLS
-Application-Layer Protocol Negotiation (ALPN) Protocol IDs".
-
-The presentation format and wire format of SvcParamValue
-is its registered "Identification Sequence".  This key SHALL NOT
-appear more than once in a SvcFieldValue.
-
-Clients MUST include this value in the ProtocolNameList in their
-ClientHello's `application_layer_protocol_negotiation` extension.
-Clients SHOULD also include any other values that they support and
-could negotiate on that connection with equivalent or better security
-properties.  For example, when using a SvcFieldValue with an "alpn" of
-"h2", the client MAY also include "http/1.1" in the ProtocolNameList.
-
-Clients MUST ignore SVCB RRs where the "alpn" SvcParamValue
-is unknown or not supported for use with the current scheme.
-
-The value of the "alpn" SvcParamKey can have effects beyond the content
-of the TLS handshake and stream.  For example, an "alpn" value of "h3"
-({{HTTP3}} Section 11.1) indicates the client must use QUIC, not TLS.
+## "transport" and "no-default-transport"
+
+The "transport" and "no-default-transport" SvcParamKeys together
+indicate the set of transport protocols supported by this service endpoint.
+
+Each scheme that is mapped to SVCB defines a set or registry of allowed
+transport values, and a "default set" of supported values, which SHOULD NOT
+be empty.  To determine the set of transport protocols supported by an
+endpoint (the "transport set"), the client collects the set of
+"transport" values, and then adds the default set unless the
+"no-default-transport" SvcParamKey is present.  The presence of a value in
+the transport set indicates that this service endpoint, described by
+SvcDomainName and the other parameters (e.g. "port") offers service with
+that transport.
```

Comment:
> This design more-or-less forces all new UDP-based protocols for transporting HTTP to share protocol invariants with QUIC.

I'm not quite sure what this means, or if it's true.  There may indeed be two UDP transports that can't share a port.  That's fine; they can just go in separate RRs with different port numbers.

We could certainly restructure this to require one transport per RR, or to express multiple ports per RR (TCP and UDP or per-transport), but the former seems inefficient, and the latter seems overcomplicated.  Sharing a port number between TCP and UDP may not be precisely principled, but in practice IANA [almost always](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers) assigns both ports to the same service, so I think it's fairly intuitive.

As you noted, sharing an RR is mostly a space optimization, especially with ESNIConfig.  My guess is that the majority of IPs that have QUIC and TLS with the same ESNIConfig will also have them at the same port number, so one-port-number-per-RR gets us most of the space savings with the least complexity.

Reply:
Sorry, I should have elaborated on "more-or-less". Encoding separate RRs for port purposes in this proposal is fussy (you need a no-default-transport) and costs a duplicate ESNI config. (Which we presumably care about or we'd go with the flattened design.) That means there will be pressure at the protocol design phase to share invariants, which means we effectively are saying this.

That's not to say this is fatal. Maybe each scheme only needs two sets of protocol invariants? QUIC folks have no doubt thought about this more than me so I'd like to know what they think. My main position is that, if we're strongly incentivizing this, we need to recognize that and decide we're okay with it, and not hide behind the verbose alternative existing.

Reply:
> Encoding separate RRs for port purposes in this proposal is fussy (you need a no-default-transport) and costs a duplicate ESNI config. (Which we presumably care about or we'd go with the flattened design.)

Agreed, although this syntax was also motivated by simplicity for zone owners when QUIC and TLS share all parameters, which we took to be the common case.

> That means there will be pressure at the protocol design phase to share invariants, which means we effectively are saying this.

I think this is at most a weak incentive, not a significant pressure on future designs.  Suppose NOTQUIC is a secure transport protocol over UDP that cannot share a port with QUIC.  (I think this is an unlikely scenario but we should nonetheless support it.)  In the current design, this would be written as
```
... port=1234 transport=quic esniconfig="abc..." ; TLS and QUIC
... port=4321 transport=notquic no-default-transport esniconfig="def..." ; NOTQUIC
```
In a flattened design, this would instead be
```
... port=1234 esniconfig="abc..." ; TLS
... port=1234 transport=quic esniconfig="abc..." ; QUIC
... port=4321 transport=notquic esniconfig="def..." ; NOTQUIC
```
I don't see a major advantage in comprehensibility between these options.  On balance, I prefer the former; duplicating the ESNIConfig seems to invite typos and misconfigurations that will not easily be caught.

BTW, even if NOTQUIC can't share a UDP port with QUIC, users can still avoid `no-default-transport` by running TLS on the TCP port of the same number.  This seems like a reasonable, perhaps likely configuration, for the same reasons that it appears to be popular with QUIC.

Reply:
The problem with the half-flattened scheme is this design is extremely inconsistent in its principles. It simultaneously believes `no-default-transport` is a rare misfeature only to be used for questionable deployment patterns, at the same time it is a pattern for new protocols. It then simultaneously believes duplicating the ESNIConfig is a problem and should be avoided, yet recommends it as a pattern for new protocols. All this mixup then pressures new protocols to share UDP ports to avoid all these design flaws, yet it also insists this pressure doesn't exist.

> BTW, even if NOTQUIC can't share a UDP port with QUIC, users can still avoid no-default-transport by running TLS on the TCP port of the same number. This seems like a reasonable, perhaps likely configuration, for the same reasons that it appears to be popular with QUIC.

No, this is not a reasonable or likely configuration. Moreover, it's dangerous because it further incentivizes existing TCP HTTP servers to not check port numbers against the HTTP `Host` header. (This check is a security requirement for sites running different origins on different ports as it's the only thing preventing the network from mixing ports up.)

It's popular with QUIC because, barring any reason to pick another port, 443 is the natural one in order to match with the origin. HTTPSSVC having broken all these standard assumption is not going to suddenly make people want to deploy TCP over the corresponding port for NOTQUIC. That doesn't really make any sense at all.

Reply:
It sounds like you're proposing to drop the requirement that zone authors need to take an explicit action to disable TLS support, as in your flattened example in this thread.  If you're willing to drop that requirement, I agree that we can find a more principled design.

Personally, I would lean toward more multiplexing, e.g.
```
... ports=tls:1234,quic:1234 esniconfig="abc..."
... ports=notquic:4321 esniconfig="def..."
```

Reply:
I'm not a fan of multiplexing based on port and making it primary in the model.  There may be transports in the future that move away from numeric ports or have a different model that isn't port-based.  As an example, if we had a future L4 protocol assignment for some QUIC successor (rather than requiring it to be over UDP).

Reply:
I think extension to portless protocols would work fine in this model.  It would just require defining one or more new SvcParamKeys.  For example, we could have
```
... ports=tls:443,quic:443,quic:8443 l4quic esniconfig="abc..."
```
where `l4quic` is a new empty-valued SvcParamKey indicating support for the (hypothetical) QUIC-at-L4 protocol.

Basically, ServiceForm is extensible enough that we can probably convey whatever information we need to in the future.

Reply:
> It sounds like you're proposing to drop the requirement that zone authors need to take an explicit action to disable TLS support, as in your flattened example in this thread.

I mean, I never requested that property to begin with. :-) I think something got lost in translation here.

The deployability requirement is that is *impossible* to disable the default transport. Possible with explicit action doesn't solve anything. It should either be syntactically impossible (preferred) or, if that can't be avoided, it needs to be easy to describe and detect (that's why I've been trying to avoid giving it a name that differs by scheme), with the expectation that authoritative servers reject it at the config level and clients reject it if they see it.

From there, it would be nice if the wrong thing was easy to avoid, but if authoritatives perform the mandatory checks, that helps smooth things over a bit.

Let's sync after the long weekend. GitHub PRs aren't a great medium for design discussions. I think we need something more synchronous and high-bandwidth. This DNS record has gotten far too complicated. We need to cut down the flexibility by an order of magnitude before this is viable.